## Data loading + class balance check

In [2]:
import pandas as pd
import numpy as np

DATA_PATH = "../data/WA_Fn-UseC_-Telco-Customer-Churn.csv"

df = pd.read_csv(DATA_PATH)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [6]:
numeric_charges = pd.to_numeric(df['TotalCharges'], errors='coerce')
blank_mask = numeric_charges.isnull()

n_blanks = blank_mask.sum()
tenure_where_blank = df.loc[blank_mask, 'tenure']
all_zero = tenure_where_blank.eq(0).all()

raw_counts = df['Churn'].value_counts()
churn_rate = df['Churn'].value_counts(normalize=True)

print(f"Number of blank TotalCharges: {n_blanks}")
print(f"All blank TotalCharges have tenure of 0: {all_zero}")
print(f"Raw counts of Churn: {raw_counts.to_dict()}")
print(f"Churn rate: {churn_rate.to_dict()}")

Number of blank TotalCharges: 11
All blank TotalCharges have tenure of 0: True
Raw counts of Churn: {'No': 5174, 'Yes': 1869}
Churn rate: {'No': 0.7346301292063041, 'Yes': 0.2653698707936959}


**Findings:** 11 blank `TotalCharges` values, all at `tenure == 0` - not missing data, just unbilled new customers. `Churn` is 73.5% No / 26.5% Yes. Seemingly imbalanced.